### DISCUSSION TITLE
What features exactly does uqbar provide?

### DISCUSSION THREAD

**Discussion by jgarte** [Abjad/abjad#1494] [Q&A] [answered]
@trevorbaca @josiah-wolf-oberholtzer 

Just for the record and for documentation, what functionality does uqbar provide for abjad?

https://github.com/josiah-wolf-oberholtzer/uqbar

The README states "Tools for creating Sphinx and Graphviz documentation." but that doesn't explain/detail what particular tools are provided.

I'll try to do some source code reading soon to try to figure it out but it would be nice to provide mid level detailed explanations of what uqbar does for general documentation purposes.

**Comment by josiah-wolf-oberholtzer** [comment, marked as answer]
Most of the functionality in Uqbar existed in Abjad in the distant past, but has been spun-off into its own package so that other libraries can use it without needing to rely on Abjad itself.

For example...

Graphs:
- Uqbar models the components of a Graphviz graph as classes, and Abjad's AbjadGrapher extends Uqbar's Grapher to configure output location.
- Anytime you call `abjad.graph`, you're hooking into Uqbar's graphing tools.

Datastructures:
- Abjad's rhythm trees (and therefore meter representations) are built on top of Uqbar's "unique tree" container classes, which are similar to Abjad's components in that each object in the tree can appear only once.

Docs:
- Abjad's API is auto-generated by the uqbar.sphinx.api extension (could potentially be replaced with https://www.sphinx-doc.org/en/master/tutorial/automatic-doc-generation.html but will be a lot of work)
- All LilyPond notation in the docs is generated and captured into the docs build process by uqbar.sphinx.book, which executes every codeblock that ends up in the documentation (this is the great grandchild of the ancient abjad-book tool that existed when I first joined the project).
